cep-20.1.md

```diff
@@ -0,0 +1,286 @@
+  - sel(unix and cmp(python, "3")):
+      url: http://path/to/python3/unix/source
+```
+
+Lists are automatically "merged" upwards, so it is possible to group multiple
+items under a single selector:
+
+```yaml
+test:
+  commands:
+    - sel(unix):
+      - test -d ${PREFIX}/include/xtensor
+      - test -f ${PREFIX}/include/xtensor/xarray.hpp
+      - test -f ${PREFIX}/lib/cmake/xtensor/xtensorConfig.cmake
+      - test -f ${PREFIX}/lib/cmake/xtensor/xtensorConfigVersion.cmake
+    - sel(win):
+      - if not exist %LIBRARY_PREFIX%\include\xtensor\xarray.hpp (exit 1)
+      - if not exist %LIBRARY_PREFIX%\lib\cmake\xtensor\xtensorConfig.cmake (exit 1)
+      - if not exist %LIBRARY_PREFIX%\lib\cmake\xtensor\xtensorConfigVersion.cmake (exit 1)
+
+# On unix this is rendered to:
+test:
+  commands:
+    - test -d ${PREFIX}/include/xtensor
+    - test -f ${PREFIX}/include/xtensor/xarray.hpp
+    - test -f ${PREFIX}/lib/cmake/xtensor/xtensorConfig.cmake
+    - test -f ${PREFIX}/lib/cmake/xtensor/xtensorConfigVersion.cmake
+```
+
+## Templating with Jinja
+
+The spec supports simple Jinja templating in the `recipe.yaml` file.
+
+You can set up Jinja variables in the context YAML section:
+
+```yaml
+context:
+  name: "test"
+  version: "5.1.2"
+  major_version: "{{ version.split('.')[0] }}"
+```
+
+Later in your `recipe.yaml` you can use these values in string interpolation
+with Jinja. For example:
+
+```yaml
+source:
+  url: https://github.com/mamba-org/{{ name }}/v{{ version }}.tar.gz
+```
+
+Jinja has built-in support for some common string manipulations.
+
+In the new spec, complex Jinja is completely disallowed as we try to produce
+YAML that is valid at all times. So you should not use any `{% if ... %}` or
+similar Jinja constructs that produce invalid yaml. Furthermore, quotes need to
+be applied when starting a value with double-curly brackets like so:
+
+```yaml
+package:
+  name: {{ name }}   # WRONG: invalid yaml
+  name: "{{ name }}" # correct
+```
+
+Some Jinja functions remain valid, but this is out of the scope of this spec.
+However, as an example, the `compiler` Jinja function will still be available,
+with the main difference being the quoting of the brackets.
+
+```yaml
+requirements:
+  build:
+    - "{{ compiler('cxx') }}"
+```
+
+## Shortcomings
+
+Since we deliberately limit the amount of "Jinja" that is allowed in recipes
+there will be several shortcomings.
+
+For example, using a `{% for ... %}` loop is prohibited. After searching through
+the conda-forge recipes with the Github search, we found for loops mostly used
+in tests. 
+
+In our view, for loops are a nice helper, but not necessary for many tasks: the
+same functionality could be achieved in a testing script, for example. At the
+same time we also plan to formalize a more powerful testing harness ([prototyped
+in
+boa](https://github.com/mamba-org/boa/blob/main/docs/source/recipe_spec.md#check-for-file-existence-in-the-final-package)).
```

Comment:
I think this looks very nice, but it's critically missing a way to check for the _absence_ of files (which is crucial for many recipes, especially multi-output ones).

It would actually be a nice improvement over what conda currently does to check for absence of files _in the output under testing_, rather than just looking at the path in the prefix. That's because a given path might be satisfied -- intentionally -- by another output; for example, it's currently not possible to check[^1] that in a lib + bin + bindings setup, the binary output (that's depending on the lib) does not accidentally package any library artefacts.

[^1]: without an extra output

Reply:
I just came up with a syntax to test the absence (negation) of the statement by using the `~` prefix. This shoudl be fairly straightforward to add. And as you suggested, these functions would operate on the `paths.json` _of the package_ (and not the environment).

But on the other hand this is a bit off topic since this is not the right CEP for that :)

Reply:
btw. conda doesn't do anything -- what people do is write ugly bash scripts that need special knowledge to handle Windows vs. Unix ... which is why testing is generally pretty bad on conda-forge. Another thing that we want to fix :) (but off-topic).

Reply:
What about the [gitignore syntax](https://git-scm.com/docs/gitignore#_pattern_format) (e.g. `!` for negation)? Devs are already familiar with that.

Reply:
https://stackoverflow.com/a/9703472

Reply:
I'm happy if there's a way to test absence of files, syntax is secondary to me. That said:

> YAML is a complex spec

I'm not deeply enough in this whole discussion, but I'd be thrilled to move to TOML as well. One thing that scares me about the `sel` / `if` discussion is that YAML breaks horrifically if you get the indentation levels wrong within a list.

Reply:
Since many folks are already coming from Python, think significant whitespace is already an accepted thing

Not to say moving to a simpler markup language might be beneficial for us, but that seems like a separate discussion altogether. Right now Wolf is trying to just move use from 2 meta languages and 1 markup language to just 1 markup language, which would be a huge improvement over the status quo. Maybe a future discussion could be moving from 1 markup language to another (once we are fully in 1 :)